### ISSUE TITLE
New zsh session throws "fasd: function definition file not found"

### ISSUE THREAD

**Issue by yeongsheng-tan** [zimfw/zimfw#232]
Please check the existing issues to make sure you're not duplicating a report.

For bug reports, please provide the following information:

In a terminal, run `zmanage info` and paste the output below:

```
Zim commit ref:   abbeeb9
Zsh version:      zsh 5.4.2 (x86_64-apple-darwin16.7.0)
System info:      Darwin YS-MacBook-Pro.local 17.2.0 Darwin Kernel Version 17.2.0: Fri Sep 29 18:27:05 PDT 2017; root:xnu-4570.20.62~3/RELEASE_X86_64 x86_64
```

- gist of .zimrc:
```
#################
# CORE SETTINGS #
#################
#
# Zim settings
#
# Select what modules you would like enabled.
# The second line of modules may depend on options set by modules in the first line.
# These dependencies are noted on the respective module's README.md.
zmodules=(directory environment git git-info archive debug ssh history input utility meta custom \
          syntax-highlighting history-substring-search prompt spectrum completion fasd)

###################
# MODULE SETTINGS #
###################

#
# Prompt
#

# Set your desired prompt here
zprompt_theme='gitster'

#
# Completion
#

# set an optional host-specific filename for the completion cache file
# if none is provided, the default '.zcompdump' is used.
zcompdump_file=".zcompdump-${HOST}-${ZSH_VERSION}"

#
# Utility
#

# Uncomment to enable command correction prompts
# See: http://zsh.sourceforge.net/Doc/Release/Options.html#Input_002fOutput
setopt CORRECT

#
# Input
#

# Uncomment to enable double-dot expansion.
# This appends '../' to your input for each '.' you type after an initial '..'
zdouble_dot_expand='true'

#
# Syntax-Highlighting
#

# This determines what highlighters will be used with the syntax-highlighting module.
# Documentation of the highlighters can be found here:
#   https://github.com/zsh-users/zsh-syntax-highlighting/blob/master/docs/highlighters.md
# For (u)rxvt, termite and gnome-terminal users,
# removing the 'cursor' highlighter will fix the disappearing cursor problem
zhighlighters=(main brackets cursor)


#
# SSH
#

# Load these ssh identities with the ssh module
zssh_ids=(id_rsa)

#
# Pacman
#

# Set (optional) pacman front-end.
#zpacman_frontend='powerpill'

# Load any helper scripts as defined here
#zpacman_helper=(aur)

# Customize to your needs...
bindkey -s '^l' 'clear\n'
# tmux
alias tmux='tmux -2'
alias tma='tmux attach -t -d'
alias tmn='tmux attach -t $(basename $(pwd)) || tmux new -s $(basename $(pwd))'
alias tml='tmux list-sessions'
alias zshconf="vim ~/.zshrc"
alias la="ls -la"
alias less=$PAGER
alias zless=$PAGER
alias ccat='pygmentize -g $1'
alias fuck='eval $(thefuck $(fc -ln -1 | tail -n 1)); fc -R'
alias gr='[ ! -z `git rev-parse --show-cdup` ] && cd `git rev-parse --show-cdup || pwd`'
alias ems="emacs --daemon"
# # IP addresses
# alias ip="dig +short myip.opendns.com @resolver1.opendns.com"
alias localip="ipconfig getifaddr en0"

export LC_ALL="en_US.UTF-8"
# Add homebrew to the completion path
# enable colored output from ls, etc
export CLICOLOR=1
# expand functions in the prompt
setopt prompt_subst
# why would you type 'cd dir' if you could just type 'dir'?
setopt AUTO_CD
# Now we can pipe to multiple outputs!
setopt MULTIOS

# This makes cd=pushd
setopt AUTO_PUSHD

# This will use named dirs when possible
setopt AUTO_NAME_DIRS

# If we have a glob this will expand it
setopt GLOB_COMPLETE
setopt PUSHD_MINUS

# No more annoying pushd messages...
setopt PUSHD_SILENT

# blank pushd goes to home
setopt PUSHD_TO_HOME

# this will ignore multiple directories for the stack. Useful? I dunno.
setopt PUSHD_IGNORE_DUPS

# 10 second wait if you do something that will delete everything. I wish I'd had this before...
setopt RM_STAR_WAIT

# use magic (this is default, but it can't hurt!)
setopt ZLE

setopt NO_HUP

# only fools wouldn't do this ;-)
# export EDITOR="subl -n -w"
setopt IGNORE_EOF

# If I could disable Ctrl-s completely I would!
setopt NO_FLOW_CONTROL

# Keep echo "station" > station from clobbering station
setopt NO_CLOBBER

# Case insensitive globbing
setopt NO_CASE_GLOB

# Be Reasonable!
setopt NUMERIC_GLOB_SORT

# I don't know why I never set this before.
setopt EXTENDED_GLOB

# hows about arrays be awesome? (that is, frew${cool}frew has frew surrounding all the variables, not just first and last
setopt RC_EXPAND_PARAM

# Who doesn't want home and end to work?
bindkey '\e[1~' beginning-of-line
bindkey '\e[4~' end-of-line

# Additional zsh plugins
test -e "${HOME}/.iterm2_shell_integration.zsh" && source "${HOME}/.iterm2_shell_integration.zsh"

# zsh-completions
fpath=(/usr/local/share/zsh-completions $fpath)

# zsh-syntax-highlighting
source /usr/local/share/zsh-syntax-highlighting/zsh-syntax-highlighting.zsh

# zsh-autosuggestions
source /usr/local/share/zsh-autosuggestions/zsh-autosuggestions.zsh

# zsh-navigation-tools
fpath+=( /usr/local/share/zsh-navigation-tools )
source /usr/local/share/zsh-navigation-tools/zsh-navigation-tools.plugin.zsh

# zsh gulp autocompletion
# source /Users/yeongsheng/.gulp-autocompletion-zsh/gulp-autocompletion.zsh

alias naliases=n-aliases ncd=n-cd nenv=n-env nfunctions=n-functions nhistory=n-history
alias nkill=n-kill noptions=n-options npanelize=n-panelize
autoload znt-history-widget
zle -N znt-history-widget
bindkey "^R" znt-history-widget
zle -N znt-cd-widget
bindkey "^B" znt-cd-widget
zle -N znt-kill-widget
bindkey "^Y" znt-kill-widget

# aws-cli zsh completion
# source /usr/local/share/zsh/site-functions/_aws

# zsh-bd
source ${HOME}/.zsh/plugins/bd/bd.zsh

# fasd
alias a='fasd -a'        # any
alias s='fasd -si'       # show / search / select
alias d='fasd -d'        # directory
alias f='fasd -f'        # file
alias sd='fasd -sid'     # interactive directory selection
alias sf='fasd -sif'     # interactive file selection
alias z='fasd_cd -d'     # cd, same functionality as j in autojump
alias zz='fasd_cd -d -i' # cd with interactive selection
```

- gist of .zshrc:
```
#
# User configuration sourced by interactive shells
#

# Source zim
if [[ -s ${ZDOTDIR:-${HOME}}/.zim/init.zsh ]]; then
  source ${ZDOTDIR:-${HOME}}/.zim/init.zsh
fi
# The following lines were added by compinstall
zstyle :compinstall filename '/Users/yeongsheng/.zshrc'
fpath=("/usr/local/bin/" $fpath)

export TERM=xterm-256color

autoload -Uz compinit
compinit
# End of lines added by compinstall
zstyle :compinstall filename '${HOME}/.zshrc'

fasd_cache="${HOME}/.fasd-init-zsh"
if [ "$(command -v fasd)" -nt "${fasd_cache}" -o ! -s "${fasd_cache}" ]; then
  fasd --init posix-alias posix-hook zsh-hook zsh-ccomp zsh-ccomp-install zsh-wcomp zsh-wcomp-install >| "${fasd_cache}"
fi
source "${fasd_cache}"
unset fasd_cache

# SCALA_HOME
export SCALA_HOME=/usr/local/opt/scala/idea

# DOCKER HOST
export DOCKER_HOST="tcp://172.30.65.153:4243"

# Add /usr/sbin for new location of mosquitto bin
export PATH=${PATH}:/usr/local/sbin

# Add imagemagick to PATH
# export PATH=${PATH}:/usr/local/opt/imagemagick/bin
export PATH="/usr/local/opt/imagemagick@6/bin:$PATH"
export PKG_CONFIG_PATH=/usr/local/opt/imagemagick/lib/pkgconfig

# Activate erlang@20 erlang shell history
export ERL_AFLAGS="-kernel shell_history enabled"

# yarn
export PATH="${HOME}/.yarn/bin:${PATH}"

# Rust/racer/rustfmt
export PATH="${HOME}/.cargo/bin:${PATH}"

# chruby
# source /usr/local/chruby/share/chruby/chruby.sh
source /usr/local/share/chruby/chruby.sh
source /usr/local/share/chruby/auto.sh
# source "$HOME/.rvm/scripts/extras/chruby.sh"
# source /Users/yeongsheng/.gulp-autocompletion-zsh/gulp-autocompletion.zsh

# zsh-auto-completion
source ~/.zsh/zsh-autosuggestions/zsh-autosuggestions.zsh
# zsh-syntax-highlighting
export ZSH_HIGHLIGHT_HIGHLIGHTERS_DIR=/usr/local/share/zsh-syntax-highlighting/highlighters
source /usr/local/share/zsh-syntax-highlighting/zsh-syntax-highlighting.zsh

# Increase ulimit for mongodb, redis and riak
ulimit -n 65536 200000

test -e "${HOME}/.iterm2_shell_integration.zsh" && source "${HOME}/.iterm2_shell_integration.zsh"

# emacs-client with elixir 1.5.x oepn command
export ELIXIR_EDITOR="emacsclient-elixir +__LINE__ __FILE__"

# fzf via Homebrew
# fzf via local installation
_has() {
  type "$1" &> /dev/null
}

if [ -e ~/.fzf ]; then
  _append_to_path ~/.fzf/bin
  source ~/.fzf/shell/key-bindings.zsh
  source ~/.fzf/shell/completion.zsh
fi

# fzf + ag configuration
if _has fzf && _has ag; then
  export FZF_DEFAULT_COMMAND='ag --nocolor -g ""'
  export FZF_CTRL_T_COMMAND="$FZF_DEFAULT_COMMAND"
  export FZF_ALT_C_COMMAND="$FZF_DEFAULT_COMMAND"
  export FZF_DEFAULT_OPTS='
  --color fg:242,bg:236,hl:65,fg+:15,bg+:239,hl+:108
  --color info:108,prompt:109,spinner:108,pointer:168,marker:168
  '
fi

[ -f ~/.fzf.zsh ] && source ~/.fzf.zsh
# Options to fzf command
export FZF_COMPLETION_OPTS='+c -x'
```

Description
-----------
1. Each new zsh session throws below warning/error:
```
${HOME}/.zim/modules/fasd/init.zsh:1: fasd: function definition file not found
```

Steps to reproduce
------------------
1. Install zim (git hash abbeeb9d375560b0e10c782e64970f22ff0dce11)
2. Enable zmodule 'fasd'
3. Launch a new zsh session

Images or other information
---------------------------
1. Nil.

**Comment by ericbn** [comment]
I got the error on the first shell session after adding `fasd` to `zmodules`, but not on subsequent ones. This may be because of how we zcompile the functions in `login_init.zsh`, that is sourced by `.zlogin`.

@yeongsheng-tan, make sure you've updated your `.zlogin` file to match the [latest zlogin template](https://github.com/Eriner/zim/blob/a8244a555a62f04f77c4056672c91df15103597b/templates/zlogin).

Still, this may be one of the only scenarios where a function zcompiled by `login_init.zsh` (sourced by `.zlogin) `is being called by a module initialized by `init.zsh` (sourced by `.zshrc)`. The `.zshrc` is executed before the `.zlogin` (see http://zsh.sourceforge.net/Intro/intro_3.html), hence the error.

TL;DR We are zcompiling code after initialization

@Eriner, thoughts? (Maybe we should be zcompiling code before initialization)

**Comment by yeongsheng-tan** [comment]
Hi @ericbn ,
thank you for the tip.

My ~/.zlogin looks like:
```

#
# startup file read in interactive login shells
#
# The following code helps us by optimizing the existing framework.
# This includes zcompile, zcompdump, etc.
#

(
  # Function to determine the need of a zcompile. If the .zwc file
  # does not exist, or the base file is newer, we need to compile.
  # These jobs are asynchronous, and will not impact the interactive shell
  zcompare() {
    if [[ -s ${1} && ( ! -s ${1}.zwc || ${1} -nt ${1}.zwc) ]]; then
      zcompile ${1}
    fi
  }

  zim_mods=${ZDOTDIR:-${HOME}}/.zim/modules
  setopt EXTENDED_GLOB

  # zcompile the completion cache; siginificant speedup.
  for file in ${ZDOTDIR:-${HOME}}/.zcomp^(*.zwc)(.); do
    zcompare ${file}
  done

  # zcompile .zshrc
  zcompare ${ZDOTDIR:-${HOME}}/.zshrc

  # zcompile some light module init scripts
  zcompare ${zim_mods}/git/init.zsh
  zcompare ${zim_mods}/utility/init.zsh
  zcompare ${zim_mods}/pacman/init.zsh
  zcompare ${zim_mods}/spectrum/init.zsh
  zcompare ${zim_mods}/completion/init.zsh
  zcompare ${zim_mods}/fasd/init.zsh

  # zcompile all .zsh files in the custom module
  for file in ${zim_mods}/custom/**/^(README.md|*.zwc)(.); do
    zcompare ${file}
  done

  # zcompile all autoloaded functions
  for file in ${zim_mods}/**/functions/^(*.zwc)(.); do
    zcompare ${file}
  done

  # syntax-highlighting
  for file in ${zim_mods}/syntax-highlighting/external/highlighters/**/*.zsh; do
    zcompare ${file}
  done
  zcompare ${zim_mods}/syntax-highlighting/external/zsh-syntax-highlighting.zsh
  
  # zsh-histery-substring-search
  zcompare ${zim_mods}/history-substring-search/external/zsh-history-substring-search.zsh

) &!
```

Do I just stick the [latest zlogin template](https://github.com/Eriner/zim/blob/a8244a555a62f04f77c4056672c91df15103597b/templates/zlogin) at the top of the line above?

**Comment by yeongsheng-tan** [comment]
@ericbn I went ahead and stuck the [latest zlogin template](https://github.com/Eriner/zim/blob/a8244a555a62f04f77c4056672c91df15103597b/templates/zlogin) to the top of my ~/.zlogin and my first fresh shell session threw the below:
```
/Users/yeongsheng/.zim/modules/fasd/init.zsh:1: fasd: function definition file not found
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/custom/init.zsh.zwc
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/spectrum/init.zsh.zwc
➜ ~ zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/fasd/init.zsh.zwc
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/prompt/functions/prompt_liquidprompt_setup.zwc
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/syntax-highlighting/external/highlighters/brackets/brackets-highlighter.zsh.zwc
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/syntax-highlighting/external/highlighters/cursor/cursor-highlighter.zsh.zwc
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/syntax-highlighting/external/highlighters/line/line-highlighter.zsh.zwc
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/syntax-highlighting/external/highlighters/main/main-highlighter.zsh.zwc
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/syntax-highlighting/external/highlighters/pattern/pattern-highlighter.zsh.zwc
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/syntax-highlighting/external/highlighters/root/root-highlighter.zsh.zwc
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.4.1): /Users/yeongsheng/.zim/modules/syntax-highlighting/external/zsh-syntax-highlighting.zsh.zwc
zrecompile:zcompile:99: zwc file has wrong version (zsh-5.3.1): /Users/yeongsheng/.zim/modules/history-substring-search/external/zsh-history-substring-search.zsh.zwc
```

Subsequent sessions came up clean and I no longer have the above errors/warnings, nor am I anymore plagued by the original issue:
```
${HOME}/.zim/modules/fasd/init.zsh:1: fasd: function definition file not found
```
👍 💯 😍

**Comment by yeongsheng-tan** [comment]
@ericbn could `zmanage update` have taken care of updating ~/.zlogin appropriately?

**Comment by ericbn** [comment]
@yeongsheng-tan, you can replace the whole content of your `.zlogin` with the [contents of the new template](https://github.com/Eriner/zim/blob/a8244a555a62f04f77c4056672c91df15103597b/templates/zlogin) (instead of prepending them), since all you have is zim's previous code there.

The main reason we did these changes are to allow next updates to be taken care of by `zmanage update`.

Still, there's the `function definition file not found` error in the first shell session after cache is cleaned that needs to be addressed, so let's keep this issue open.
